File: server/models/WorkingHours.js
const mongoose = require("mongoose");

const workingHoursSchema = new mongoose.Schema(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    date: {
      type: Date,
      required: true,
    },
    targetHours: {
      type: Number,
      required: true,
      default: 15,
    },
    achievedHours: {
      type: Number,
      required: true,
      default: 0,
    },
    category: {
      type: String,
      required: true,
    },
    notes: {
      type: String,
      trim: true,
    },
    mood: {
      type: String,
      enum: ["Productive", "Normal", "Distracted"],
      default: "Normal",
    },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true },
  }
);

// Virtual for progress percentage
workingHoursSchema.virtual("progressPercentage").get(function () {
  return (this.achievedHours / this.targetHours) * 100;
});

// Virtual for completion status
workingHoursSchema.virtual("status").get(function () {
  const percentage = this.progressPercentage;
  if (percentage >= 100) return "Completed";
  if (percentage >= 75) return "On Track";
  if (percentage >= 50) return "In Progress";
  return "Behind";
});

// Compound index to prevent duplicate entries for the same user and date
workingHoursSchema.index({ user: 1, date: 1 }, { unique: true });

module.exports = mongoose.model("WorkingHours", workingHoursSchema);

File: backend-go/internal/models/working_hours.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type WorkingHours struct {
	ID            primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	User          primitive.ObjectID `bson:"user" json:"user"`
	Date          time.Time          `bson:"date" json:"date"`
	TargetHours   float64            `bson:"targetHours" json:"targetHours"`
	AchievedHours float64            `bson:"achievedHours" json:"achievedHours"`
	Category      string             `bson:"category" json:"category"`
	Notes         string             `bson:"notes,omitempty" json:"notes,omitempty"`
	Mood          string             `bson:"mood" json:"mood"`
	CreatedAt     time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt     time.Time          `bson:"updatedAt" json:"updatedAt"`
}

// Virtual properties logic (ProgressPercentage, Status) will be handled in the Controller/Service layer or a method.
func (w *WorkingHours) ProgressPercentage() float64 {
	if w.TargetHours == 0 {
		return 0
	}
	return (w.AchievedHours / w.TargetHours) * 100
}

func (w *WorkingHours) Status() string {
	percentage := w.ProgressPercentage()
	if percentage >= 100 {
		return "Completed"
	}
	if percentage >= 75 {
		return "On Track"
	}
	if percentage >= 50 {
		return "In Progress"
	}
	return "Behind"
}
